### ISSUE TITLE
Dense Captioning Evaluation on VG Dataset

### ISSUE THREAD

**Issue by WaikeiGum** [JialianW/GRiT#6]
Hello,

I am currently tring to reproduce the result of task dense captioning of GRiT. I have trained the model by default setting and got the checkpoint of it. Then I ran inference on VG test set and got the json result by 

`python train_net.py --num-gpus-per-machine 8 --config-file configs/GRiT_B_DenseCap.yaml --output-dir-name ./output/grit_b_densecap --eval-only MODEL.WEIGHTS models/grit_b_densecap.pth`

However, when installing the environment of DenseCap, I was stuck in the installation of torch on my GPU machine which has a CUDA version of 12.0. I always met this error:

Make Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDA_cublas_device_LIBRARY (ADVANCED)
    linked by target "THC" in directory /root/torch/extra/cutorch/lib/THC

Could you tell me what platform you use to install DenseCap and perform evaluation?

Thanks a lot!

**Comment by JialianW** [comment]
To evaluate the result, you can use the docker as provided in https://github.com/jcjohnson/densecap/issues/95

**Comment by WaikeiGum** [comment]
Hi, Jialian. 

I have built the environment for densecap using the docker image as provided. Could you please tell me more specifically how you use the output/grit_b_densecap/vg_instances_results.json or checkpoint file to evaluate on it so as to get the mAP result?

Found an evaluator of VG in python 
https://github.com/soloist97/densecap-pytorch/blob/bc81d9816ff8d4e45613846ad2acdf789acde37b/model/evaluator.py#L72

**Comment by JialianW** [comment]
In their Lua evaluation code, there is a place that obtains model inference results. We replace it with our result read from json.

**Comment by WaikeiGum** [comment]
> In their Lua evaluation code, there is a place that obtains model inference results. We replace it with our result read from json.

Thanks a lot.

**Comment by WaikeiGum** [comment]
Hi, Jialian!

Have you ever met this problem : "attempt to concatenate a nil value" in eval_utils.lua ?  

I met this when evaluating the model on image 63.jpg   [while reading ground truth](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/eval/eval_utils.lua#L63)

It seems likes that it is a problem caused by [nn.LanguageModel.idx_to_token](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/densecap/LanguageModel.lua#L98)   when the idx=10579, the token is "nil", which will cause error when being concatenated.

**Comment by JialianW** [comment]
I only changed the code to read the results json file, and no other changes. I didn't have errors in the evaluation code.

**Comment by WaikeiGum** [comment]
Could you please provide me the evaluation code you edit so i could replace it in my docker?  Thanks

**Comment by EvenYYY** [comment]
Hi, Jialian.

I found that the vocabulary size of VG is 10509 which takes single word as a token, while GRiT takes a subword as a token and thus gets a vacabulary size of 30522. So that is why i got the error above.

**Comment by amsword** [comment]
> Hi, Jialian!
> 
> Have you ever met this problem : "attempt to concatenate a nil value" in eval_utils.lua ?
> 
> I met this when evaluating the model on image 63.jpg [while reading ground truth](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/eval/eval_utils.lua#L63)
> 
> It seems likes that it is a problem caused by [nn.LanguageModel.idx_to_token](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/densecap/LanguageModel.lua#L98) when the idx=10579, the token is "nil", which will cause error when being concatenated.

Can you show more details on how you hit the issue? For example, share the full error stack.

**Comment by EvenYYY** [comment]
Yes, thanks for your reply. I found that the vocab size is only 10509 in densecap after preprocessing. And so it raises error when idx = 10579, that is because in preprocesing, densecap itself filter a great deal of words the shows less frequently, i.e. [less than 15 times](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/preprocess.py#L462). Have you change this number?

**Comment by EvenYYY** [comment]
When doing evaluation, in VG it takes all the result of an image as an input to addResult(), do you combine all the boxes, region_descriptions and confidences of boxes in a single image as one input to addResult()?

**Comment by WaikeiGum** [comment]
> > Hi, Jialian!
> > Have you ever met this problem : "attempt to concatenate a nil value" in eval_utils.lua ?
> > I met this when evaluating the model on image 63.jpg [while reading ground truth](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/eval/eval_utils.lua#L63)
> > It seems likes that it is a problem caused by [nn.LanguageModel.idx_to_token](https://github.com/jcjohnson/densecap/blob/7c32170f134805debe638806ecb0a22bbcd58c5f/densecap/LanguageModel.lua#L98) when the idx=10579, the token is "nil", which will cause error when being concatenated.
> 
> Can you show more details on how you hit the issue? For example, share the full error stack.

I found the error: I didn't replace the vocab_size and idx_to_token in LM model, thank you all so much.

**Comment by hellowordo** [comment]
@EvenYYY 您好，请问一下您可以提供一下您编辑的用vg_instances_results.json文件在DenseCap评估的代码嘛，非常感谢您！两位大佬上面的聊天我还是没看太懂
